### PR TITLE
Implemented Feature to Prevent Negative Review Spamming by Disgruntled Tenants

### PR DESCRIPTION
**Pull request by jallen2034**
I implemented a new feature int he Nest backend codebase to prevent Tenants from posting negative review spam and also refactored the review service a bit to be cleaner and more modular.

### CODE REVIEW

File: backend/.vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        "localRoot": "${workspaceFolder}",
+        "remoteRoot": "/app",
+        "restart": true,
+        "sourceMaps": true,
+        "outFiles": [
+          "${workspaceFolder}/dist/**/*.js"
+        ]
+      }
+    ]
+  }
```
Comment (jallen2034) [inline backend/.vscode/launch.json:18]: Added launch.json to configure a remote debugger to the NestJS backend in VS code. 

I can remove this file from the commit if requested :)

File: backend/package.json [hunk 2]
```diff
@@ -11,7 +11,7 @@
     "format": "prettier --write \"src/**/*.ts\" \"test/**/*.ts\"",
     "start": "nest start",
     "start:dev": "nest start --watch",
-    "start:debug": "nest start --debug --watch",
+    "start:debug": "nest start --debug=0.0.0.0:9999 --watch",
```
Comment (jallen2034) [inline backend/package.json:14]: Opened up a remote debug port on the NestJS backend running out of the docker container locally.

File: backend/src/captcha/captcha-service.ts [hunk 3]
```diff
@@ -1,13 +1,14 @@
...
+import { BadRequestException } from '@nestjs/common';
 
 @Injectable()
 export class CaptchaService {
   VERIFY_URL = 'https://hcaptcha.com/siteverify';
 
   constructor(private readonly httpService: HttpService) {}
 
-  async verifyToken(token: string, ip?: string): Promise<boolean> {
+  async verifyToken(token: string, ip?: string): Promise<any> {
```
Comment (jallen2034) [inline backend/src/captcha/captcha-service.ts:11]: Don't recall actually touching this, not sure how it changed :/

File: backend/src/review/review.service.spec.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+      const result = await service.checkRreviewsForSimilarity(
+        reviewsFromDb,
+        reviewUserSubmitted,
+      );
+
+      expect(result).toBe(false);
+    });
+  });
+});
```
Comment (jallen2034) [inline backend/src/review/review.service.spec.ts:138]: Working unit tests for my new ReviewSimilarityService who's purpose is to check if the review text coming back from the Database is too similar to the review text a person submits for a landlord in a new review.

Comment (jallen2034) [reply backend/src/review/review.service.spec.ts:138]: image

File: backend/src/review/review-text-match.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        review.review,
+        reviewUserSubmitted,
+      );
+      if (similarityScore > 0.7) {
+        return true;
+      }
+    }
+    return false;
+  }
+}
```
Comment (jallen2034) [inline backend/src/review/review-text-match.ts:70]: New Class/service with the methods and logic to handle detecting if a user is likely spamming reviews against a landlord.

File: backend/src/review/review.service.ts [hunk 6]
```diff
@@ -16,9 +17,27 @@ type ReviewQuery = {
...
+      return await this.databaseService.sql<Review[]>`SELECT REVIEW
+        FROM review
+        WHERE landlord = ${inputReview.landlord}
+          AND ZIP = ${inputReview.zip};`;
+    } catch (e) {
+      throw new InternalServerErrorException(
+        'Failed to retrieve existing reviews from the database',
+      );
+    }
+  }
```
Comment (jallen2034) [inline backend/src/review/review.service.ts:38]: Get existing reviews for the landlord who's review is about to be submitted

File: backend/src/captcha/captcha-service.ts (hunk 3 above)
Comment (kellenwiltshire) [reply backend/src/captcha/captcha-service.ts:11]: Will need to update this to get rid of the `any`. Should go back to `boolean`

File: backend/src/review/review.controller.ts [hunk 7]
```diff
@@ -18,13 +20,30 @@ import { IStats, Review, ReviewsResponse } from './models/review';
...
+};
+
 @Controller('review')
 export class ReviewController {
   constructor(
     private readonly reviewService: ReviewService,
     private captchaService: CaptchaService,
   ) {}
 
+  private handleException(e: any): never {
```
Comment (kellenwiltshire) [inline backend/src/review/review.controller.ts:35]: No `any`

**Review by kellenwiltshire — CHANGES_REQUESTED** [review]
Couple changes. Still need to test this out though myself

File: backend/src/review/review.service.ts [hunk 8]
```diff
@@ -16,9 +17,27 @@ type ReviewQuery = {
...
+    private reviewSimilarityService: ReviewSimilarityService,
+  ) {}
+
+  private async getExistingRevewsForUser(
+    inputReview: Review,
+  ): Promise<Review[]> {
+    try {
+      return await this.databaseService.sql<Review[]>`SELECT REVIEW
+        FROM review
+        WHERE landlord = ${inputReview.landlord}
```
Comment (kellenwiltshire) [inline backend/src/review/review.service.ts:31]: Make sure to normalize the input here. Should be `${inputReview.landlord.toLocaleUpperCase()}`

File: backend/src/review/review.service.ts [hunk 9]
```diff
@@ -16,9 +17,27 @@ type ReviewQuery = {
...
+          AND ZIP = ${inputReview.zip};`;
```
Comment (kellenwiltshire) [inline backend/src/review/review.service.ts:32]: This line as well, should be `AND zip = ${inputReview.zip.toLocaleUpperCase()}`

**Review by kellenwiltshire — CHANGES_REQUESTED** [review]
Couple things to fix up. Otherwise things look good!

File: backend/src/review/review.service.ts (hunk 9 above)
Comment (jallen2034) [reply backend/src/review/review.service.ts:32]: image

Resolved :)

File: backend/src/review/review.service.ts [hunk 10]
```diff
@@ -107,164 +149,223 @@ export class ReviewService {
...
   }
 
-  findOne(id: number): Promise<Review[]> {
-    return this.databaseService.sql<
-      Review[]
-    >`Select * FROM review WHERE id IN(${id});`;
+  public findOne(id: number): Promise<Review[]> {
+    return this.databaseService.sql<Review[]>`Select *
+      FROM review
+      WHERE id IN (${id});`;
```
Comment (jallen2034) [inline backend/src/review/review.service.ts:155]: Think this just got auto formatted by my IDE when saving the file.

File: backend/src/review/review.controller.ts (hunk 7 above)
Comment (jallen2034) [reply backend/src/review/review.controller.ts:35]: image

Fixed :)

File: backend/src/captcha/captcha-types.ts [hunk 11]
```diff
@@ -0,0 +1,6 @@
+export type CaptchaPayload = {
+  secret: string;
+  sitekey: string;
+  response: string;
+  remoteip: string;
+};
```
Comment (jallen2034) [inline backend/src/captcha/captcha-types.ts:6]: More TypeScript documentation added.

File: backend/src/review/models/review-data-layer.ts [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+        `
+      )[0].id;
+
+      inputReview.id = id;
+      return inputReview;
+    } catch (e) {
+      throw e;
+    }
+  }
+}
```
Comment (jallen2034) [inline backend/src/review/models/review-data-layer.ts:46]: ## ReviewModel - Refactored SQL Query and Future Improvements

I have refactored the code in `ReviewModel` to enhance adherence to the MVC architecture by moving the SQL query for creating a review into this data layer class. This separation of concerns improves code organization and maintainability.

In the future, we can further improve the backend code of the review service by migrating other SQL queries into methods within `ReviewModel`. This approach will help to clean up the codebase and provide a more structured implementation :)

File: backend/src/review/review.service.ts [hunk 13]
```diff
@@ -16,9 +19,26 @@ type ReviewQuery = {
 
 @Injectable()
 export class ReviewService {
-  constructor(private readonly databaseService: DatabaseService) {}
+  constructor(
+    private readonly databaseService: DatabaseService,
+    private reviewSimilarityService: ReviewSimilarityService,
+    private reviewDataLayerService: ReviewModel,
+  ) {}
```
Comment (jallen2034) [inline backend/src/review/review.service.ts:26]: I have enhanced the `ReviewService` class by introducing dependency injections for better code organization and modularity.

The following dependencies have been injected:
- `databaseService`: Allows access to the database functionality.
- `reviewSimilarityService`: Provides the ability to perform review similarity calculations.
- `reviewDataLayerService`: Offers methods to interact with the database for review-related operations.

By injecting these dependencies, we achieve a decoupled and more maintainable codebase. It also paves the way for further extensibility and testability.

Please review the changes and provide any feedback or suggestions you may have.

File: backend/src/review/review.service.ts [hunk 14]
```diff
@@ -107,164 +150,195 @@ export class ReviewService {
...
+        await this.reviewSimilarityService.checkRreviewsForSimilarity(
+          existingReviews,
+          inputReview.review,
+        );
+
+      if (reviewSpamDetected) return inputReview;
+      return this.reviewDataLayerService.createReview(inputReview, filterResult); // hit data layer to create review
+    } catch (e) {
+      throw e;
+    }
```
Comment (jallen2034) [inline backend/src/review/review.service.ts:175]: The create method in the ReviewService has been significantly improved, resulting in a cleaner and more maintainable codebase. By extracting the SQL logic into the createReview method of the ReviewDataLayerService class, we have achieved better separation of concerns.

Previously, the create method handled both the AI filtering and the creation of the review in the database. However, with the refactoring, the AI filtering is now handled separately, and the database interaction is delegated to the ReviewDataLayerService.

This separation allows for clearer code organization and enhances the readability and testability of the create method. It also adheres to the principles of the Single Responsibility Principle (SRP) and the Separation of Concerns (SoC) design principles.

**Review by kellenwiltshire — APPROVED** [review]
Looks good!

**Pull request merged**